packages/plugin-list/List.ts

```diff
@@ -23,80 +24,135 @@ function _isEmptyTextNode(node: Node): boolean {
+            // blocks.
             return [context, parsingMap];
         }
-        // Inline elements in a list item should be wrapped in a paragraph.
-        if (!utils.isBlock(children[0]) || children[0].nodeName === 'BR') {
-            const paragraph = new ParagraphNode(); // todo: remove reference to plugin
-            context.parentVNode.append(paragraph);
-            context.parentVNode = paragraph;
-            parsingMap.set(paragraph, [context.currentNode]);
+    }
+    /**
+     * Render the ListNode in currentContext.
+     */
+    async renderToDom(node: VNode, domParent: Node): Promise<Map<VNode, Node[]>> {
+        if (node.is(ListNode)) {
+            const tag = node.listType === ListType.ORDERED ? 'OL' : 'UL';
+            const domListNode = document.createElement(tag);
+            domParent.appendChild(domListNode);
+
+            let mapping: [VNode, Node[]][] = [[node, [domListNode]]];
+
+            // Render the node's children
+            for (const child of node.children) {
+                const childrenMapping = await this.renderListItemToDom(child, domListNode);
+                mapping = [...mapping, ...childrenMapping];
+            }
+
+            return new Map(mapping);
         }
-        // Now we can move on to the list item's contents, to be added to
-        // the paragraph created above, or to the list itself in the case of
-        // blocks.
-        return [context, parsingMap];
     }
+    async renderListItemToDom(node: VNode, domParent: Node): Promise<Map<VNode, Node[]>> {
+        // The ListNode has to handle the rendering of its direct children by
+        // itself since some of them are rendered inside "LI" nodes while others
+        // are rendered *as* "LI" nodes.
+        // Check if previous "LI" can be reused or create a new one.
+        let domListItem: Element;
+        if (node.is(ListNode) && domParent.childNodes.length) {
+            // Render an indented list in the list item that precedes it.
+            // eg.: <ul><li>title: <ul><li>indented</li></ul></ul>
+            domListItem = domParent.childNodes[domParent.childNodes.length - 1] as Element;
+        } else {
+            domListItem = document.createElement('li');
+        }
+        domParent.appendChild(domListItem);
 
-    //--------------------------------------------------------------------------
-    // Public
-    //--------------------------------------------------------------------------
+        // Direct ListNode's VElement children "P" are rendered as "LI"
+        // while other nodes will be rendered inside the "LI".
+        if (node.is(VElement) && node.htmlTag === 'P') {
+            await this.editor.renderers.dom.renderChildren(node, domListItem);
+        } else {
+            // The node was wrapped in a "LI" but needs to be rendered as well.
+            await this.editor.renderers.dom.renderNode(node, domListItem);
```

Comment:
The `renderNode` function is on the renderer, right ? What if it was this function which does the mapping ? It would map the result of the rendering function to the `VNode` it got in argument. No more need for manual mappings *unless* you create unrepresented nodes. Example:
```typescript
// In DomRenderer.ts
renderNode(node: VNode) {
    const result = renderingFunction(node, this.domMap);
    if (result) {
        this.domMap.set(node, result);
    }
}

// In List.ts
renderListToDom(node: ListNode, domMap: Map<VNode: Node[]>) {
    const ul = createUL();
    for (const child of node) {
        const li = createLI();
        const domChild = this.editor.render(child);
        li.appendChild(domChild);
        domMap.get(child).unshift(li);
        ul.appendChild(li);
    }
}
```

Reply:
Thanks for using ```typescript.